### DISCUSSION TITLE
A way to pass "self" parameter without reference on C imported methods.

### DISCUSSION THREAD

**Discussion by Andre-LA** [edubart/nelua-lang#9] [Q&A] [answered]
Hi
It is not something urgent, but it would be nice to have a way to specify the _self_ argument in a C imported method call.

_Context:_

There is two external .h files, [one with this struct](https://github.com/raysan5/raylib/blob/39e73ccc4d1a09c5276b2a575a65081002a1143c/src/raylib.h#L171) and [another with this function](https://github.com/raysan5/raylib/blob/39e73ccc4d1a09c5276b2a575a65081002a1143c/src/raymath.h#L176):
```c
typedef struct Vector2 { float x; float y; } Vector2;// from raylib.h
Vector2 Vector2Add(Vector2 v1, Vector2 v2); // from raymath.h
```

In my .nelua file, I imported the struct in this way:
```lua
global Vector2 <cimport, nodecl> = @record{x: float32, y: float32}
```

then, I imported the ``Vector2Add`` function on a separated record since raymath is an external module separated from the main one:

```lua
global Raymath = @record{}
function Raymath.Vector2Add(v1: Vector2, v2: Vector2): Vector2 <cimport'Vector2Add', nodecl> end
```

but, is would be nice to have an Add method on Vector2 too, so I tried the following:

```lua
function Vector2:Add(v2: Vector2): Vector2 <cimport'Vector2Add', nodecl> end

-- test:
local x: Vector2 = { 0, 10 }
local x2: Vector2 = { 10, 0 }
x:Add(x2)
```

however, it's rendered as:

```c
// [...]
// I abbreviated the names
static Vector2 x = {0, 10};
static Vector2 x2 = {10, 0};
// [...]
// this fails because Vector2Add expects an ``Vector2``
Vector2Add(&x, x2);
```

**Comment by edubart** [comment, marked as answer]
When using `:`, the self argument is always declared as a pointer to the object, so this code:
```lua
function Vector2:Add(v2: Vector2): Vector2 <cimport'Vector2Add', nodecl> end
```
is really a shortcut for:
```lua
function Vector2.Add(self: *Vector2, v2: Vector2): Vector2 <cimport'Vector2Add', nodecl> end
```

As you can see is wrong, because the imported C function doesn't expect a pointer in the `v1` argument.
In this case you need to explicitly tell the first argument, so like in lua use `.` instead of `:`, do this:

```lua
function Vector2.Add(v1: Vector2, v2: Vector2): Vector2 <cimport'Vector2Add', nodecl> end
```
then the `x:Add(x2)` should work as expected.

**Plus TIP:** On the very last commit I've implemented binary operator metamethods for records so now you should be able to do:

```lua
function Vector2.__add(v1: Vector2, v2: Vector2): Vector2 <cimport'Vector2Add', nodecl> end
```

then this should work:
```lua
local x: Vector2 = { 0, 10 }
local y: Vector2 = { 10, 0 }
local z = x + y
```

The same metamethods of Lua5.3 is available for overloading, that is:

```
__eq       a == b
__lt       a < b
__le       a <= b
__bor      a | b
__bxor     a ~ b
__band     a & b
__shl      a << b
__shr      a >> b
__concat   a .. b
__add      a + b
__sub      a - b
__mul      a * b
__idiv     a // b
__div      a / b
__pow      a ^ b
__mod      a % b
__len      #a
__unm      -a
```

**Comment by Andre-LA** [comment]
Thanks!
